design/ceph/osd-migration.md

```diff
@@ -0,0 +1,188 @@
+
+- Add a new label `osd-store:<osd store type>` to all the OSD pods.
+- This label will help to identify the current backend store being used for the OSD and enable the operator to decide if the OSD should be migrated in case the user changes the backend store type in the spec (`spec.Storage.store.type`).
+
+### New OSDs
+
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the OSDs.
+
+#### OSD Prepare
+
+- The `ceph-volume prepare` command is updated to use backend store provided in the `spec.storage.store`. If no value is provided for `spec.storage.store` then default to `bluestore` backend.
+
+    - RAW MODE
+        ```
+        ceph-volume raw prepare <OSD_STORE_FLAG> --data /dev/vda
+        ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm prepare <OSD_STORE_FLAG> --data /dev/vda
+      ```
+
+#### OSD Create
+
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.store`. If no values are provided by the user, then `OSD_STORE_FLAG` defaults to `bluestore`
+- `ceph-volume` commands use the `OSD_STORE_FLAG` while activating the OSDs.
+
+    - RAW MODE
+      ```
+      ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs
+      ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"
+      ```
+
+**NOTE**: ceph-volume `activate` command doesn't require `OSD_STORE_FLAG` to be passed as an argument. It auto detects the backend store that was used during the OSD `prepare`
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a Ceph version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  store:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### Replace OSD
+- Operator reconciler will replace one OSD at a time.
+- A config map will be used to store the list of OSDs to be replaced and the OSD being replaced currently.
+- OSD replacement steps:
+  1. List out all OSDs where `osd-store:<osd store type>` does not match with `spec.storage.store.type`.
+  2. These OSDs will not be upgraded before migrating to the new OSD store. This is required because any upgrade operation will update the osd store to `bluestore-rdr` in the OSD deployment spec without migrating the OSD. As a result, the operator will lose track of the OSDs that need migration.
```

Comment:
In a previous discussion, we decided to always preserve the store type label on the OSD after an OSD has been created, right? Then we can avoid losing track of which OSDs to migrate because we expect the OSDs to have the correct label.

Reply:
I missed that [comment](https://github.com/rook/rook/pull/12381#discussion_r1279592535).  This can be good idea to implement. But I can't think how we can use a single osd-store label  and migrate the OSDs while preserving that label value.

Reply:
still checking more on this.

Reply:
There is way. We don't need to preserve the osd labels as well. OSD prepare pod prepares the OSD and stores the OSDInfo in the config map. This OSDInfo has osd details including the osdStore type. This OSDInfo is used by OSD deployment pod to create the OSDs. 

Currently the OSDStore type is being read from the [environment variable](https://github.com/rook/rook/blob/7aaf46e11243b21df4d4f088fc80b949019559b0/pkg/daemon/ceph/osd/volume.go#L863).  We can change this approach and use the actual osd store form ceph-volume output. That is, 

``` Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log raw list --format json
2023-08-07 04:48:45.870980 D | cephosd: {
    "0efd66fa-f4e3-47c6-a7d7-cc0b5d5c7144": {
        "ceph_fsid": "53be30b3-5bc8-495b-8eff-8c670e4b7902",
        "device": "/dev/vde",
        "osd_id": 5,
        "osd_uuid": "0efd66fa-f4e3-47c6-a7d7-cc0b5d5c7144",
        "type": "bluestore"
    },
```

This will allow OSDs (yet to be migrated to new store) to upgrade without changing the `osd-store` label on the OSD deployment spec. So users can stop migration at any point of time.

Reply:
Removed this section from the design: 
```   2. These OSDs will not be upgraded before migrating to the new OSD store. This is required because any upgrade operation will update the osd store to `bluestore-rdr` in the OSD deployment spec without migrating the OSD. As a result, the operator will lose track of the OSDs that need migration.```